controllers/substitution.go

```diff
@@ -0,0 +1,42 @@
+package controllers
+
+import (
+	"strings"
+
+	v1 "k8s.io/api/apps/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	// commonCloudControllerNamingSubstringPattern contains a substring which all CCM controller container names should have
+	// to be substituted with image from config
+	commonCloudControllerNamingSubstringPattern = "cloud-controller-manager"
+)
+
+// setDeploymentImages substitutes controller containers in Deployment with correct image
+func setDeploymentImages(config operatorConfig, d *v1.Deployment) {
+	for i, container := range d.Spec.Template.Spec.Containers {
+		if !strings.Contains(container.Name, commonCloudControllerNamingSubstringPattern) {
+			continue
+		}
+
+		d.Spec.Template.Spec.Containers[i].Image = config.ControllerImage
+	}
+}
+
+func fillConfigValues(config operatorConfig, templates []client.Object) []client.Object {
+	objects := make([]client.Object, len(templates))
+	for i, objectTemplate := range templates {
+		// Set namespaces for all object. Namespace on cluster-wide objects is stripped by API server and is not applied
+		objectTemplate.SetNamespace(config.ManagedNamespace)
+
+		dep, ok := objectTemplate.(*v1.Deployment)
+		if ok {
+			setDeploymentImages(config, dep)
+			// TODO: add cloud-config calculated hash to annotations to account for redeployment on content change
+		}
+
+		objects[i] = objectTemplate
+	}
```

Comment:
This is ripe for having leaky side effects, we should have copies of the templates, not modify them in place directly
```suggestion
	for i, objectTemplate := range templates {
		obj, ok := objectTemplate.DeepCopy().(client.Object)
		if !ok {
		  // This should never happen. DeepCopy returns the same underlying
		  // type so it will always implement the interface.
		  panic("cloned object was not a client.Object")
		}
		// Set namespaces for all object. Namespace on cluster-wide objects is stripped by API server and is not applied
		obj.SetNamespace(config.ManagedNamespace)

		dep, ok := objectTemplate.(*v1.Deployment)
		if ok {
			setDeploymentImages(config, dep)
			// TODO: add cloud-config calculated hash to annotations to account for redeployment on content change
		}

		objects[i] = obj
	}
```